underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java

```diff
@@ -489,42 +490,38 @@ private boolean isRoot(String key) {
       path = stripPrefixIfPresent(path);
       path = PathUtils.normalizePath(path, PATH_SEPARATOR);
       path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
+      // If non-recursive, let the listObjects only get the files in the folder
+      String delimiter = recursive ? null : PATH_SEPARATOR;
+
+      // NOTE(binfan): currently OSS client only supports listing at most 1000 objects, without
+      // setting continuation
       ListObjectsRequest listObjectsRequest = new ListObjectsRequest(mBucketName);
       listObjectsRequest.setPrefix(path);
+      listObjectsRequest.setMaxKeys(1000);
+      listObjectsRequest.setDelimiter(delimiter);
 
-      // recursive, so don't set the delimiter
-      // then list will return all files in this dir and subdirs
-      if (recursive) {
-        ObjectListing listing = mClient.listObjects(listObjectsRequest);
-        List<OSSObjectSummary> objectSummaryList = listing.getObjectSummaries();
-        String[] ret = new String[objectSummaryList.size()];
-        for (int i = 0; i < objectSummaryList.size(); i++) {
-          // Remove parent portion of the key
-          String child = getChildName(objectSummaryList.get(i).getKey(), path);
-          // Prune the special folder suffix
-          child = stripFolderSuffixIfPresent(child);
-          ret[i] = child;
-        }
-        return ret;
-      }
-
-      // Non recursive, so set the delimiter, let the listObjects only get the files in the folder
-      listObjectsRequest.setDelimiter(PATH_SEPARATOR);
-      Set<String> children = new HashSet<String>();
+      Set<String> children = new HashSet<>();
```

Comment:
Good catch. Actually, the others should use `Set` too like this as duplicated children are possible.
Updated